WebTransport should send network errors to JS for failure points

AnkshitJain opened the pull request:
#### d09b3b34e3e94ff6842abb400d692f156173edfb
<pre>
WebTransport should send network errors to JS for failure points
<a href="https://bugs.webkit.org/show_bug.cgi?id=285087">https://bugs.webkit.org/show_bug.cgi?id=285087</a>
<a href="https://rdar.apple.com/136263051">rdar://136263051</a>

Reviewed by Alex Christensen.

Network connections are canceled when the session is terminated.
Streams handle completion and error for reads and writes.
Incoming connections are canceled if the incoming streams reader is closed.

* Source/WebCore/Modules/streams/ReadableStreamSource.h:
* Source/WebCore/Modules/webtransport/DatagramSink.cpp:
(WebCore::DatagramSink::write):
(WebCore::DatagramSink::send):
* Source/WebCore/Modules/webtransport/DatagramSink.h:
* Source/WebCore/Modules/webtransport/DatagramSource.cpp:
(WebCore::DatagramSource::receiveDatagram):
(WebCore::DatagramSource::doCancel):
* Source/WebCore/Modules/webtransport/DatagramSource.h:
* Source/WebCore/Modules/webtransport/WebTransport.cpp:
(WebCore::WebTransport::receiveDatagram):
(WebCore::WebTransport::receiveIncomingUnidirectionalStream):
(WebCore::WebTransport::receiveBidirectionalStream):
* Source/WebCore/Modules/webtransport/WebTransport.h:
* Source/WebCore/Modules/webtransport/WebTransportBidirectionalStreamSource.cpp:
(WebCore::WebTransportBidirectionalStreamSource::receiveIncomingStream):
* Source/WebCore/Modules/webtransport/WebTransportBidirectionalStreamSource.h:
* Source/WebCore/Modules/webtransport/WebTransportReceiveStreamSource.cpp:
(WebCore::WebTransportReceiveStreamSource::receiveIncomingStream):
* Source/WebCore/Modules/webtransport/WebTransportReceiveStreamSource.h:
* Source/WebCore/Modules/webtransport/WebTransportSession.h:
* Source/WebCore/Modules/webtransport/WebTransportSessionClient.h:
* Source/WebCore/PAL/pal/cocoa/NetworkSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/NetworkSoftLink.mm:
* Source/WebCore/PAL/pal/spi/cocoa/NetworkSPI.h:
* Source/WebCore/bindings/js/ReadableStreamDefaultController.h:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.cpp:
(WebKit::NetworkTransportSession::sendDatagram):
(WebKit::NetworkTransportSession::streamSendBytes):
(WebKit::NetworkTransportSession::terminate):
(WebKit::NetworkTransportSession::receiveDatagram):
(WebKit::NetworkTransportSession::streamReceiveBytes):
(WebKit::NetworkTransportSession::receiveIncomingUnidirectionalStream):
(WebKit::NetworkTransportSession::receiveBidirectionalStream):
(WebKit::NetworkTransportSession::cancelReceiveStream):
(WebKit::NetworkTransportSession::cancelSendStream):
(WebKit::NetworkTransportSession::destroyStream):
(WebKit::NetworkTransportSession::sendStreamSendBytes): Deleted.
(WebKit::NetworkTransportSession::destroyOutgoingUnidirectionalStream): Deleted.
(WebKit::NetworkTransportSession::destroyBidirectionalStream): Deleted.
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.h:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.messages.in:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportStream.cpp:
(WebKit::NetworkTransportStream::NetworkTransportStream):
(WebKit::NetworkTransportStream::sendBytes):
(WebKit::NetworkTransportStream::cancelReceive):
(WebKit::NetworkTransportStream::cancelSend):
(WebKit::NetworkTransportStream::cancel):
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportStream.h:
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
(WebKit::NetworkTransportSession::initialize):
(WebKit::NetworkTransportSession::setupDatagramConnection):
(WebKit::NetworkTransportSession::sendDatagram):
(WebKit::NetworkTransportSession::setupConnectionHandler):
(WebKit::NetworkTransportSession::createStream):
(WebKit::NetworkTransportSession::receiveDatagramLoop):
(WebKit::NetworkTransportSession::terminate):
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportStreamCocoa.mm:
(WebKit::NetworkTransportStream::NetworkTransportStream):
(WebKit::NetworkTransportStream::sendBytes):
(WebKit::NetworkTransportStream::receiveLoop):
(WebKit::NetworkTransportStream::setErrorCodeForStream):
(WebKit::NetworkTransportStream::cancel):
(WebKit::NetworkTransportStream::cancelReceive):
(WebKit::NetworkTransportStream::cancelSend):
* Source/WebKit/WebProcess/Network/WebTransportReceiveStreamSource.cpp:
(WebKit::WebTransportReceiveStreamSource::receiveBytes):
(WebKit::WebTransportReceiveStreamSource::doCancel):
* Source/WebKit/WebProcess/Network/WebTransportReceiveStreamSource.h:
* Source/WebKit/WebProcess/Network/WebTransportSendStreamSink.cpp:
(WebKit::WebTransportSendStreamSink::write):
(WebKit::WebTransportSendStreamSink::close):
(WebKit::WebTransportSendStreamSink::error):
(WebKit::WebTransportSendStreamSink::sendBytes):
* Source/WebKit/WebProcess/Network/WebTransportSendStreamSink.h:
* Source/WebKit/WebProcess/Network/WebTransportSession.cpp:
(WebKit::WebTransportSession::receiveDatagram):
(WebKit::WebTransportSession::receiveIncomingUnidirectionalStream):
(WebKit::WebTransportSession::receiveBidirectionalStream):
(WebKit::WebTransportSession::streamReceiveBytes):
(WebKit::WebTransportSession::sendDatagram):
(WebKit::WebTransportSession::streamSendBytes):
(WebKit::WebTransportSession::cancelReceiveStream):
(WebKit::WebTransportSession::cancelSendStream):
(WebKit::WebTransportSession::destroyStream):
* Source/WebKit/WebProcess/Network/WebTransportSession.h:
* Source/WebKit/WebProcess/Network/WebTransportSession.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebTransport.mm:
(TestWebKitAPI::TEST(WebTransport, DISABLED_ClientBidirectional)):
(TestWebKitAPI::TEST(WebTransport, DISABLED_Datagram)):
(TestWebKitAPI::TEST(WebTransport, DISABLED_Unidirectional)):
(TestWebKitAPI::TEST(WebTransport, DISABLED_ServerBidirectional)):

Canonical link: <a href="https://commits.webkit.org/288498@main">https://commits.webkit.org/288498@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3f9a601adfcaf6ee992a50d0ceca3edd2b808a5e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/83621 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/3238 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/37921 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/88692 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/34629 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/85706 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/3326 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/11196 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/65044 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/22787 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/86667 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/2443 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/75978 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/45331 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/2355 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/30188 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/33677 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/73423 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/30930 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/90069 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/10886 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/7855 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/73477 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/11109 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/71803 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/72703 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/16954 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/15659 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/2210 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12902 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/10838 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/16310 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/10686 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/14161 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/12458 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
```diff
@@ -366,6 +378,7 @@ static void didReceiveServerTrustChallenge(Ref<NetworkConnectionToWebProcess>&&
         ASSERT(!strongThis->m_streams.contains(identifier));
         strongThis->m_streams.set(identifier, stream.releaseNonNull());
         completionHandler(identifier);
+        completionHandler = nullptr;
```
This is not necessary.  A CompletionHandler becomes null when it is called.

achristensen07 commented on Source/WebKit/WebProcess/Network/WebTransportReceiveStreamSource.cpp:
```diff
@@ -44,21 +44,38 @@ WebTransportReceiveStreamSource::~WebTransportReceiveStreamSource()
+        return;
+    if (exception) {
+        controller().error(*exception);
+        clean();
```
I'm not very familiar with these calls.  Do you know what happens when clean is not called?

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
```diff
@@ -263,17 +277,21 @@ static void didReceiveServerTrustChallenge(Ref<NetworkConnectionToWebProcess>&&
+            if (!(nw_error_get_error_domain(error) == nw_error_domain_posix && nw_error_get_error_code(error) == ECANCELED))
+                completionHandler(WebCore::Exception(WebCore::ExceptionCode::NetworkError));
+            // If we were cancelled, exit gracefully
+            return;
```
If it's not a posix cancelled error, we won't ever call the completionHandler, which will assert.  It looks like we're missing a completionHandler(std::nullopt)

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
```diff
@@ -405,23 +415,53 @@ static void didReceiveServerTrustChallenge(Ref<NetworkConnectionToWebProcess>&&
+{
+#if HAVE(WEB_TRANSPORT)
+    RetainPtr metadata = adoptNS((nw_protocol_metadata_t)nw_retain(nw_connection_group_copy_protocol_metadata(m_connectionGroup.get(), nw_protocol_copy_webtransport_definition())));
+    if (metadata) {
```
if (RetainPtr metadata = adoptNS(...))
Then metadata has the right scope and you don't need to explicitly set it to null.  Also, the nw_retain call looks like it'll cause a memory leak.

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportStreamCocoa.mm:
```diff
@@ -41,46 +47,164 @@
+            if (!(nw_error_get_error_domain(error) == nw_error_domain_posix && nw_error_get_error_code(error) == ECANCELED))
+                completionHandler(WebCore::Exception(WebCore::ExceptionCode::NetworkError));
+            // If we were cancelled, exit gracefully
+            return;
```
Ditto.  completionHandler should be called in all possible flows.

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportStreamCocoa.mm:
```diff
@@ -41,46 +47,164 @@
-                return true;
-            });
+            if (content) {
+                dispatch_data_apply_span(content, [&](std::span<const uint8_t> buffer) {
```
Is dispatch_data_apply_span API?  Can we just use dispatch_data_apply instead?

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportStreamCocoa.mm:
```diff
@@ -41,46 +47,164 @@
+        return;
+
+#if HAVE(WEB_TRANSPORT)
+    RetainPtr metadata = adoptNS(nw_connection_copy_protocol_metadata(m_connection.get(), nw_protocol_copy_webtransport_definition()));
```
nw_protocol_copy_webtransport_definition needs adoptNS or it will leak.

achristensen07 commented on Source/WebCore/Modules/webtransport/DatagramSource.cpp:
```diff
@@ -35,20 +35,29 @@ DatagramSource::DatagramSource() = default;
 {
     m_isCancelled = true;
 }
-
```
nit: leave this space here.

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/NetworkTransportStream.h:
```diff
@@ -53,8 +55,10 @@ class NetworkTransportStream : public RefCounted<NetworkTransportStream>, public
-    void sendBytes(std::span<const uint8_t>, bool withFin);
-
+    void sendBytes(std::span<const uint8_t>, bool, CompletionHandler<void(std::optional<WebCore::Exception>&&)>&&);
+    void cancelReceive(std::optional<unsigned>);
```
Are error codes 32 bits?  If so, I think we should do something like this:
using NetworkTransportErrorCode = uint32_t
Otherwise it's a little unclear what the unsigned is.

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportStreamCocoa.mm:
```diff
@@ -41,46 +47,166 @@
+
+#if HAVE(WEB_TRANSPORT)
+    RetainPtr metadata = adoptNS(nw_connection_copy_protocol_metadata(m_connection.get(), adoptNS(nw_protocol_copy_webtransport_definition()).get()));
+    if (metadata)
```
nit: variable declaration can be inside the if statement to match its scope.

achristensen07 commented on Source/WebKit/WebProcess/Network/WebTransportSendStreamSink.cpp:
```diff
@@ -64,10 +67,38 @@ void WebTransportSendStreamSink::write(WebCore::ScriptExecutionContext& context,
+        return;
+    RefPtr session = m_session.get();
+    if (session)
+        session->streamSendBytes(m_identifier, std::span<const uint8_t>(), true);
```
We can probably just use { } for the span.  We might want to use a 2-state enum for WithFin to make the "true" easier to tell what it means to a future reader.

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WebTransport.mm:
```diff
@@ -220,10 +229,15 @@ WebTransportServer echoServer([](ConnectionGroup group) -> Task {
         "    let {value: s, d} = await sr.read();"
         "    let r = s.readable.getReader();"
         "    const { value, done } = await r.read();"
+        "    await r.cancel();"
```
To test a lot of this I think it would be useful to make it so the server can terminate connections without having been asked to by JS.

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportStreamCocoa.mm:
```diff
@@ -41,46 +47,162 @@
+    if (!errorCode)
+        return;
+
+    // FIXME: Implement once rdar://141886375 is available in OS builds.
```
We could probably do this before then with optional soft linking.

AnkshitJain replied:
I'll add this in a separate PR.

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportStreamCocoa.mm:
```diff
@@ -41,46 +47,162 @@
+        if (error) {
+            if (!(nw_error_get_error_domain(error) == nw_error_domain_posix && nw_error_get_error_code(error) == ECANCELED))
+                session->streamReceiveBytes(strongThis->m_identifier, { }, false, WebCore::Exception(WebCore::ExceptionCode::NetworkError));
+            // If we were cancelled, exit gracefully
```
Do we want to do anything with the error if it is not ECANCELED?  This comment is a little mysterious.

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportStreamCocoa.mm:
```diff
@@ -41,46 +47,162 @@
-    ASSERT(m_streamType != NetworkTransportStreamType::OutgoingUnidirectional);
+    if (m_streamState == NetworkTransportStreamState::ReadClosed) {
+        RELEASE_ASSERT_NOT_REACHED();
+        return;
```
This is unreachable.  This whole thing could just be RELEASE_ASSERT(m_streamState != NetworkTransportStreamState::ReadClosed)

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
```diff
@@ -228,30 +231,40 @@ static void didReceiveServerTrustChallenge(Ref<NetworkConnectionToWebProcess>&&
+            return;
+        }
+        if (!session)
+            completionHandler(nullptr);
```
This needs to return or it will call the completion handler twice which will dereference null the second time because CompletionHandler sets itself to null when it is called.

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
```diff
@@ -228,30 +231,40 @@ static void didReceiveServerTrustChallenge(Ref<NetworkConnectionToWebProcess>&&
+        completionHandler = WTFMove(completionHandler)
+    ] (RefPtr<NetworkTransportSession>&& session) mutable {
+        if (!completionHandler) {
+            session = nullptr;
```
I don't think this is needed.

achristensen07 commented on Source/WebKit/WebProcess/Network/WebTransportSession.cpp:
```diff
@@ -165,4 +170,22 @@ void WebTransportSession::networkProcessCrashed()
+
+void WebTransportSession::cancelReceiveStream(WebCore::WebTransportStreamIdentifier identifier, std::optional<WebCore::WebTransportStreamErrorCode> errorCode)
+{
+    ASSERT(RunLoop::isMain());
```
I don't think this is correct.  WorkerWebTransportSession::cancelReceiveStream calls this directly from the non-main run loop.  I'd just remove these.

achristensen07 commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
```diff
@@ -405,23 +415,51 @@ static void didReceiveServerTrustChallenge(Ref<NetworkConnectionToWebProcess>&&
+
+    for (auto& stream : m_streams.values())
+        stream->cancel(code);
+    m_streams.clear();
```
Instead of this, you could iterate std::exchange(m_streams, { }).values() to do both at the same time.  Not too important though.

achristensen07 commented on Source/WebCore/Modules/webtransport/WebTransportReceiveStreamSource.cpp:
```diff
@@ -27,32 +27,72 @@
-void WebTransportReceiveStreamSource::doCancel()
+WebTransportReceiveStreamSource::WebTransportReceiveStreamSource()
+    : m_transport(nullptr)
+    , m_identifier(std::nullopt)
```
These lines don't do anything.  Let's remove them.

Commit:
WebTransport should send network errors to JS for failure points
https://bugs.webkit.org/show_bug.cgi?id=285087
rdar://136263051

Reviewed by Alex Christensen.

Network connections are canceled when the session is terminated.
Streams handle completion and error for reads and writes.
Incoming connections are canceled if the incoming streams reader is closed.

* Source/WebCore/Modules/streams/ReadableStreamSource.h:
* Source/WebCore/Modules/webtransport/DatagramSink.cpp:
(WebCore::DatagramSink::write):
(WebCore::DatagramSink::send):
* Source/WebCore/Modules/webtransport/DatagramSink.h:
* Source/WebCore/Modules/webtransport/DatagramSource.cpp:
(WebCore::DatagramSource::receiveDatagram):
(WebCore::DatagramSource::doCancel):
* Source/WebCore/Modules/webtransport/DatagramSource.h:
* Source/WebCore/Modules/webtransport/WebTransport.cpp:
(WebCore::WebTransport::receiveDatagram):
(WebCore::WebTransport::receiveIncomingUnidirectionalStream):
(WebCore::WebTransport::receiveBidirectionalStream):
* Source/WebCore/Modules/webtransport/WebTransport.h:
* Source/WebCore/Modules/webtransport/WebTransportBidirectionalStreamSource.cpp:
(WebCore::WebTransportBidirectionalStreamSource::receiveIncomingStream):
* Source/WebCore/Modules/webtransport/WebTransportBidirectionalStreamSource.h:
* Source/WebCore/Modules/webtransport/WebTransportReceiveStreamSource.cpp:
(WebCore::WebTransportReceiveStreamSource::receiveIncomingStream):
* Source/WebCore/Modules/webtransport/WebTransportReceiveStreamSource.h:
* Source/WebCore/Modules/webtransport/WebTransportSession.h:
* Source/WebCore/Modules/webtransport/WebTransportSessionClient.h:
* Source/WebCore/PAL/pal/cocoa/NetworkSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/NetworkSoftLink.mm:
* Source/WebCore/PAL/pal/spi/cocoa/NetworkSPI.h:
* Source/WebCore/bindings/js/ReadableStreamDefaultController.h:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.cpp:
(WebKit::NetworkTransportSession::sendDatagram):
(WebKit::NetworkTransportSession::streamSendBytes):
(WebKit::NetworkTransportSession::terminate):
(WebKit::NetworkTransportSession::receiveDatagram):
(WebKit::NetworkTransportSession::streamReceiveBytes):
(WebKit::NetworkTransportSession::receiveIncomingUnidirectionalStream):
(WebKit::NetworkTransportSession::receiveBidirectionalStream):
(WebKit::NetworkTransportSession::cancelReceiveStream):
(WebKit::NetworkTransportSession::cancelSendStream):
(WebKit::NetworkTransportSession::destroyStream):
(WebKit::NetworkTransportSession::sendStreamSendBytes): Deleted.
(WebKit::NetworkTransportSession::destroyOutgoingUnidirectionalStream): Deleted.
(WebKit::NetworkTransportSession::destroyBidirectionalStream): Deleted.
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.h:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportSession.messages.in:
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportStream.cpp:
(WebKit::NetworkTransportStream::NetworkTransportStream):
(WebKit::NetworkTransportStream::sendBytes):
(WebKit::NetworkTransportStream::cancelReceive):
(WebKit::NetworkTransportStream::cancelSend):
(WebKit::NetworkTransportStream::cancel):
* Source/WebKit/NetworkProcess/webtransport/NetworkTransportStream.h:
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
(WebKit::NetworkTransportSession::initialize):
(WebKit::NetworkTransportSession::setupDatagramConnection):
(WebKit::NetworkTransportSession::sendDatagram):
(WebKit::NetworkTransportSession::setupConnectionHandler):
(WebKit::NetworkTransportSession::createStream):
(WebKit::NetworkTransportSession::receiveDatagramLoop):
(WebKit::NetworkTransportSession::terminate):
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportStreamCocoa.mm:
(WebKit::NetworkTransportStream::NetworkTransportStream):
(WebKit::NetworkTransportStream::sendBytes):
(WebKit::NetworkTransportStream::receiveLoop):
(WebKit::NetworkTransportStream::setErrorCodeForStream):
(WebKit::NetworkTransportStream::cancel):
(WebKit::NetworkTransportStream::cancelReceive):
(WebKit::NetworkTransportStream::cancelSend):
* Source/WebKit/WebProcess/Network/WebTransportReceiveStreamSource.cpp:
(WebKit::WebTransportReceiveStreamSource::receiveBytes):
(WebKit::WebTransportReceiveStreamSource::doCancel):
* Source/WebKit/WebProcess/Network/WebTransportReceiveStreamSource.h:
* Source/WebKit/WebProcess/Network/WebTransportSendStreamSink.cpp:
(WebKit::WebTransportSendStreamSink::write):
(WebKit::WebTransportSendStreamSink::close):
(WebKit::WebTransportSendStreamSink::error):
(WebKit::WebTransportSendStreamSink::sendBytes):
* Source/WebKit/WebProcess/Network/WebTransportSendStreamSink.h:
* Source/WebKit/WebProcess/Network/WebTransportSession.cpp:
(WebKit::WebTransportSession::receiveDatagram):
(WebKit::WebTransportSession::receiveIncomingUnidirectionalStream):
(WebKit::WebTransportSession::receiveBidirectionalStream):
(WebKit::WebTransportSession::streamReceiveBytes):
(WebKit::WebTransportSession::sendDatagram):
(WebKit::WebTransportSession::streamSendBytes):
(WebKit::WebTransportSession::cancelReceiveStream):
(WebKit::WebTransportSession::cancelSendStream):
(WebKit::WebTransportSession::destroyStream):
* Source/WebKit/WebProcess/Network/WebTransportSession.h:
* Source/WebKit/WebProcess/Network/WebTransportSession.messages.in:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebTransport.mm:
(TestWebKitAPI::TEST(WebTransport, DISABLED_ClientBidirectional)):
(TestWebKitAPI::TEST(WebTransport, DISABLED_Datagram)):
(TestWebKitAPI::TEST(WebTransport, DISABLED_Unidirectional)):
(TestWebKitAPI::TEST(WebTransport, DISABLED_ServerBidirectional)):

Canonical link: https://commits.webkit.org/288498@main
